### PR TITLE
proxmox_backup: refactor permission checking

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY

Create a generic function to check permissions and simplify the module logic by calling it.

##### ISSUE TYPE

- Refactoring Pull Request

##### COMPONENT NAME

proxmox_backup

### CODE REVIEW

File: plugins/modules/proxmox_backup.py [hunk 1]
```diff
@@ -291,57 +294,28 @@ def check_storage_permissions(
...
-        else:
-            self.module.fail_json(
-                changed=False,
-                msg="Insufficient permission: Datastore.AllocateSpace is missing")
-        if (bandwidth or performance) and permissions["/"].get(
-                "Sys.Modify", 0) == 0:
-            self.module.fail_json(
-                changed=False,
-                msg="Insufficient permission: Performance_tweaks and bandwidth require 'Sys.Modify' permission for '/'")
+        if not has_permission(permissions, "Datastore.AllocateSpace", search_scopes=["/", "/storage", "/storage" + storage]):
```
Comment (IamLunchbox) [inline plugins/modules/proxmox_backup.py:297]: `"/storage" + storage` is missing a slash as a separator

File: plugins/modules/proxmox_backup.py [hunk 2]
```diff
@@ -291,57 +294,28 @@ def check_storage_permissions(
...
-                pass
-            elif "/storage" in permissions.keys() and permissions["/storage"].get("Datastore.Allocate", 0) == 1:
-                pass
-            elif "/storage/" + storage in permissions.keys() and permissions["/storage/" + storage].get("Datastore.Allocate", 0) == 1:
-                pass
-            else:
-                self.module.fail_json(
-                    changed=False,
-                    msg="Insufficient permissions: Custom retention was requested, but Datastore.Allocate is missing")
+            if not has_permission(permissions, "Datastore.Allocate", search_scopes=["/", "/storage", "/storage" + storage]):
```
Comment (IamLunchbox) [inline plugins/modules/proxmox_backup.py:304]: `"/storage" + storage` is missing a slash as a separator

File: plugins/modules/proxmox_backup.py [hunk 3]
```diff
@@ -351,23 +325,11 @@ def check_vmid_backup_permission(self, permissions, vmids, pool):
...
-        elif "/vms" in permissions.keys() and permissions["/vms"].get("VM.Backup", 0) == 1:
-            pass
-        elif pool and "/pool/" + pool in permissions.keys() and permissions["/pool/" + pool].get(
-                "VM.Backup", 0) == 1:
-            pass
-        else:
-            self.module.fail_json(
-                changed=False,
-                msg="Insufficient permissions: You dont have the VM.Backup permission")
+        if not has_permission(permissions, "VM.Backup", search_scopes=["/", "/vms"] + (["/vms" + pool] if pool else [])):
```
Comment (IamLunchbox) [inline plugins/modules/proxmox_backup.py:331]: search_scopes should be ` ["/", "/vms"] + (["/pool" + pool] if pool else [])`

**Comment by IamLunchbox** [comment]
Thank you for the refactor! The code is way clearer with your new function and less clutter from unnecessary list comprehensions.

Since the unit test did not catch the errors, should they be extended to test most of the different permission trees? RIght now, I went the "easy" route and just unit tested a "default" permission tree, which does not contain more complex or nested permissions and stuck to simple permissions for "/".

**Review by IamLunchbox — COMMENTED** [review]
Thank you for the refactor! The code is way clearer with your new function and less clutter from unnecessary list comprehensions.

Since the unit test did not catch the errors, should they be extended to test most of the different permission trees? RIght now, I went the "easy" route and just unit tested a "default" permission tree, which does not contain more complex or nested permissions and stuck to simple permissions for "/".

File: plugins/modules/proxmox_backup.py (hunk 1 above)
Comment (russoz) [reply plugins/modules/proxmox_backup.py:297]: Like this?
```suggestion
        if not has_permission(permissions, "Datastore.AllocateSpace", search_scopes=["/", "/storage/", "/storage" + storage]):
```

File: plugins/modules/proxmox_backup.py (hunk 3 above)
Comment (russoz) [reply plugins/modules/proxmox_backup.py:331]:
```suggestion
        if not has_permission(permissions, "VM.Backup", search_scopes=["/", "/vms"] + (["/pool" + pool] if pool else [])):
```

File: plugins/modules/proxmox_backup.py (hunk 2 above)
Comment (russoz) [reply plugins/modules/proxmox_backup.py:304]:
```suggestion
            if not has_permission(permissions, "Datastore.Allocate", search_scopes=["/", "/storage", "/storage/" + storage]):
```

File: plugins/modules/proxmox_backup.py [hunk 4]
```diff
@@ -291,57 +294,28 @@ def check_storage_permissions(
...
 
         if not sufficient_permissions:
             # Since VM.Backup can be given for each vmid at a time, iterate through all of them
             # and check, if the permission is set
             failed_vmids = []
             for vm in vmids:
-                if "/vms/" + \
-                        str(vm) in permissions.keys() and permissions["/vms/" + str(vm)].get("VM.Backup", 1) == 0:
+                vm_path = "/vms/" + str(vm)
+                if permissions.get(vm_path, {}).get("VM.Backup", 1) == 0:
```
Comment (russoz) [inline plugins/modules/proxmox_backup.py:318]:
```suggestion
                if has_permission(permissions, "VM.Backup", search_scopes=[vm_path], default=1, expected=0):
```

File: plugins/modules/proxmox_backup.py [hunk 5]
```diff
@@ -351,23 +325,11 @@ def check_vmid_backup_permission(self, permissions, vmids, pool):
...
-        elif "/vms" in permissions.keys() and permissions["/vms"].get("VM.Backup", 0) == 1:
-            pass
-        elif pool and "/pool/" + pool in permissions.keys() and permissions["/pool/" + pool].get(
-                "VM.Backup", 0) == 1:
-            pass
-        else:
-            self.module.fail_json(
-                changed=False,
-                msg="Insufficient permissions: You dont have the VM.Backup permission")
+        if not has_permission(permissions, "VM.Backup", search_scopes=["/", "/vms"] + (["/pool" + pool] if pool else [])):
```
Comment (russoz) [inline plugins/modules/proxmox_backup.py:331]: Or maybe
```suggestion
        if not has_permission(permissions, "VM.Backup", search_scopes=["/", "/vms"] + (["/pool/" + pool] if pool else [])):
```

Comment (IamLunchbox) [reply plugins/modules/proxmox_backup.py:331]: Yes, the permission tree is like a file tree, each entry is separated by a slash. 

Sorry, i just shouldve suggested the correct change. :/

Comment (russoz) [reply plugins/modules/proxmox_backup.py]: All good, thanks for keeping an eye on this. 😁

**Comment by russoz** [comment]
> Since the unit test did not catch the errors, should they be extended to test most of the different permission trees? RIght now, I went the "easy" route and just unit tested a "default" permission tree, which does not contain more complex or nested permissions and stuck to simple permissions for "/".

Not sure what you meant by "the unit test did not catch the errors". Which errors? This refactor is no bugfix, AFAICT.

**Comment by IamLunchbox** [comment]
I meant: The missing slashes couldve been caught by unit tests. 

But their location was not reached, because the test-permissions are leading the permission check to return after the first possible check. 

So I'll probably add more detailed permissions for that in the future.

**Comment by felixfontein** [comment]
What's the state of this PR?

**Comment by russoz** [comment]
AFAICT, ready for review. Unless @IamLunchbox wants to change something in it.

**Review by IamLunchbox — APPROVED** [review]
LGMT

**Comment by felixfontein** [comment]
@russoz thanks for your contribution!
@IamLunchbox thanks for reviewing!

**Pull request merged**